Symintify pad ops

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87304
* #87291
* #87231
* __->__ #87046

Following comments below, we need to add support for `std::negate`/`std::min`/`std::max`/`operator-` for SymInt

albanD committed:
symintify pad ops

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -4859,15 +4859,15 @@ Tensor sparse_constructor_values_backward(
       negated_pad.cbegin(),
       negated_pad.cend(),
       negated_pad.begin(),
-      std::negate<int64_t>());
```
We should add a std::negate specialization for SymInt. We should also make operator- work. With the branch, there isn't any particular reason to rush this

albanD committed:
Update on "symintify pad ops"




[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/PadNd.cpp:
```diff
@@ -137,63 +137,63 @@ Tensor _pad_circular(const Tensor &self, IntArrayRef padding) {
-      out_slice = out.slice(dim, out_shape[dim] - pad_r, out_shape[dim]);
-      in_slice = out.slice(dim, std::max(pad_l, zero), std::max(pad_l, zero) + pad_r);
+      out_slice = out.slice_symint(dim, out_shape[dim] - pad_r, out_shape[dim]);
+      in_slice = out.slice_symint(dim, pad_l.max(zero), pad_l.max(zero) + pad_r);
```
We should have a SymInt specialization for std::max

bdhirsh replied:
I added one last week but ended up removing it.

From the [docs](https://en.cppreference.com/w/cpp/algorithm/max), it looks like `std::max/min()` already have a template that just calls into `operator<` and `operator>`. So adding a special handling for min/max on symints in C++ seemed redundant to me, but lmk if you think that's wrong. We still have min/max support in python.

ezyang replied:
ok, I am fine with this resolution

albanD committed:
Update on "[WIP] symintify pad ops"


Following comments below, we need to add support for `std::negate`/`std::min`/`std::max`/`operator-` for SymInt

[ghstack-poisoned]

albanD committed:
Update on "[WIP] symintify pad ops"


Following comments below, we need to add support for `std::negate`/`std::min`/`std::max`/`operator-` for SymInt

[ghstack-poisoned]

albanD committed:
Update on "[WIP] symintify pad ops"


Following comments below, we need to add support for `std::negate`/`std::min`/`std::max`/`operator-` for SymInt

[ghstack-poisoned]

albanD committed:
Update on "Symintify pad ops"


Following comments below, we need to add support for `std::negate`/`std::min`/`std::max`/`operator-` for SymInt

[ghstack-poisoned]

albanD committed:
Update on "Symintify pad ops"


Following comments below, we need to add support for `std::negate`/`std::min`/`std::max`/`operator-` for SymInt

[ghstack-poisoned]

albanD committed:
Update on "Symintify pad ops"


Following comments below, we need to add support for `std::negate`/`std::min`/`std::max`/`operator-` for SymInt

[ghstack-poisoned]

albanD commented:
@ezyang this is ready for a final review

albanD committed:
Update on "Symintify pad ops"


Following comments below, we need to add support for `std::negate`/`std::min`/`std::max`/`operator-` for SymInt

[ghstack-poisoned]

albanD commented:
@pytorchbot merge -g

albanD commented:
@pytorchbot merge -g
